hail/src/main/scala/is/hail/expr/ir/NormalizeNames.scala

```diff
@@ -9,32 +9,29 @@ import scala.annotation.tailrec
 import scala.collection.mutable
 
 object NormalizeNames {
-  def apply[T <: BaseIR](
-    ctx: ExecuteContext,
-    ir: T,
-    allowFreeVariables: Boolean = false,
-  ): T = {
-    val freeVariables: Set[Name] = ir match {
-      case ir: IR =>
-        if (allowFreeVariables) {
-          val env = FreeVariables(ir, true, true)
-          env.eval.m.keySet union
-            env.agg.map(_.m.keySet).getOrElse(Set.empty) union
-            env.scan.map(_.m.keySet).getOrElse(Set.empty) union
-            env.relational.m.keySet
-        } else {
+  def apply[T <: BaseIR](ctx: ExecuteContext, ir: T, allowFreeVariables: Boolean = false): T =
+    ctx.time {
```

Comment:
My intention was to remove `time[A](context: String)(f: => A): A` altogether and use implicit information about the surrounding context instead (`sourcecode.Enclosing` for now). Sadly my dreams were foiled as it would have meant more changes to `LoweringPass` and `Optimize` which I didn't want to do; it would make an already busy change even busier. I arrived at a compromise therefore:

- In places were we specified the context as a literal, replace with the implicit context 
- Otherwise, do the old behaviour and consider a subsequent change to do the same thing.

I want to move to more implicit timing context for a couple of reasons:
1. It's less verbose which makes refactoring much much easier (subsequent change incoming)
2. You can obtain a nice timing trace that's linked to the source code and customisable

Reply:
Yeah I think so. `NormalizeNames` is called in a couple of other timed contexts like `ForwardLets` and `Compile{WithAggregators}` too. 

I don't really see a downside to always timing `NormalizeNames` internally (or indeed any larger units of named code**  (see below)). This pattern helps obtains a trace of the program which is useful in debugging. If it causes too much slow down, we could use a null-pattern for the Timer and only turn on timing if we're investigating something performance related for example.

** To your point about seeing `NormalizeNames` twice in the timer output.
Yes. That's annoying. I originally re-wrote optimise to avoid this and instrumented the various optimisations therein. I abandoned that because this change was already quite large and I didn't want to make it any spicer.

Would you be happy if I were to handle that in a subsequent change?

Reply:
Fixed in https://github.com/hail-is/hail/pull/14731